File: 2606. Find the Substring With Maximum Cost/Solution.java
class Solution {
    public int maximumCostSubstring(String s, String chars, int[] vals) {
        Map<Character, Integer> map = new HashMap<>();
        int n = vals.length, m = s.length(), res = 0, curr = 0;
        for (int i = 0; i < n; i++) map.put(chars.charAt(i), vals[i]);
        for (int i = 0; i < m; i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                curr += map.get(c);
            } else {
                curr += s.charAt(i) - 'a' + 1;
            }
            res = Math.max(res, curr);
            if (curr < 0) curr = 0;
        }
        return res;
    }
}

File: 2606. Find the Substring With Maximum Cost/Solution.py
class Solution:
    def maximumCostSubstring(self, s: str, chars: str, vals: List[int]) -> int:
        m = dict(zip(chars, vals))
        curr, res = 0, 0
        for c in s:
            if c in m:
                curr += m[c]
            else:
                curr += ord(c) - ord('a') + 1
            res = max(res, curr)
            curr = max(curr, 0)
        return res